### DISCUSSION TITLE
Asset checks on partitioned assets

### DISCUSSION THREAD

**Discussion by johannkm** [dagster-io/dagster#17194] [General]
As of Dagster 1.5.3, we support per-asset checks. We don't currently support per-partition checks ([Github issue](https://github.com/dagster-io/dagster/issues/17005)). However you can still write a check against a partitioned asset. To do so you'll need to summarize the state of all partitions in to one pass/fail for the asset.

Here's an example:

```python
from dagster import (
    asset,
    asset_check,
    Definitions,
    StaticPartitionsDefinition,
    AssetCheckResult,
)
from typing import Dict, Any
import random


@asset(
    group_name="asset_checks",
    partitions_def=StaticPartitionsDefinition(["a", "b", "c", "d"]),
)
def partitioned_asset(_):
    return random.choice(["foo", "bar"])


@asset_check(asset=partitioned_asset)
def check_on_partitioned_asset(partitioned_asset: Dict[str, Any]):
    # input is in the form of {partition_key: output}
    failed_partitions = [
        key for key, output in partitioned_asset.items() if output != "foo"
    ]
    return AssetCheckResult(
        passed=bool(failed_partitions),
        metadata={"non_foo_partitions": failed_partitions},
    )


defs = Definitions(
    assets=[partitioned_asset], asset_checks=[check_on_partitioned_asset]
)
```

The check receives a mapping of all the partition keys to the output values (the same as having an unpartitioned asset downstream of a partitioned asset). It tests each partition and fails if any don't pass.

**Comment by johannkm** [comment]
A caveat with the current approach is that if you backfill `partitioned_asset`, `check_on_partitioned_asset` will execute 4 times- once with each partition materialization. We plan to add the option to exclude checks from backfills. We may also need to add the option to only execute the check in the final backfill run, once all partitions are present.

**Comment by j-blackwell** [comment]
I might be missing something, but I don't think this works when you have an asset that uses a type handler such as pandas dataframe on a `DbIOManager` as the IO manager relies upon and checks against the type hint. So I get the following error:

```
dagster._check.CheckError: DuckDBIOManager does not have a handler for type 'typing.Dict[str, typing.Any]'. Has handlers for types '<class 'graphs.resources.query_type_handlers.SqlQuery'>, <class 'pandas.core.frame.DataFrame'>'. Please build the DuckDBIOManager with an type handler for type 'typing.Dict[str, typing.Any]', so the DuckDBIOManager can correctly handle the output.
```

**Reply by johannkm** [reply]
Hey @jrstats sorry I missed this comment. With dataframes, instead of a `partition -> value` mapping as input to the check, you'll just get a dataframe with multiple rows. Here's an example:

```python
    @asset(
        partitions_def=StaticPartitionsDefinition(["a", "b", "c"]),
        metadata={"partition_expr": "col1"},
    )
    def my_asset(context) -> pd.DataFrame:
        return pd.DataFrame({"col1": [context.partition_key], "col2": ["foo"]})

    @asset_check(asset=my_asset)
    def my_check(my_asset: pd.DataFrame):
        # this will pass once my_asset has run for all 3 partitions
        return AssetCheckResult(passed=my_asset.equals(
            pd.DataFrame({"col1": ["a", "b", "c"], "col2": ["foo", "foo", "foo"]})
        ))

```

**Reply by johannkm** [reply]
(this is the same as an unpartitioned asset downstream of a partitioned asset, it gets an all-to-one mapping)

**Reply by j-blackwell** [reply]
Thanks @johannkm , I will try this in the new year, but am I right in thinking the entire dataset (all partitions) gets loaded if you include it in the check's parameters?

**Reply by johannkm** [reply]
That's right
